### ISSUE TITLE
Compile-phase warning erroneous in required module from warning-handler

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5820]
Migrated from [rt.perl.org#16049](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=16049) (status was 'resolved')

Searchable as RT16049$

**Comment by p5pRT** [comment]
### From dj@mips.com

#### Created by dj@mips.com



\#\! /usr/local/bin/perl \-w
\#
\# The \-w flag is needed to cause the strange behavior\.
\#
\# This is a testcase for a very strange msg that perl is emitting at compile time&#8203;:
\#
\#   Can't use global @&#8203;\_ in "my" at othermod\.pm line 2\, near "return @&#8203;\_"
\#   Compilation failed in require at djstrangefailure\.pl line 27\.
\#
\# There is no "my" declaration\, so I dont know exactly what is going wrong\.
\# It seems to be confusing the parsing of the require block with that of the source code that caused the warning
\#
\# In order for this testcase to work\, you must have a file named othermod\.pm which contains
\# the following 4 lines \(uncommented\, of course\!\)&#8203;:
\#
\# sub mysub \{
\#    return @&#8203;\_;
\# \}
\# 1;
\# 

\# Register a warning subroutine for compile warnings
BEGIN \{
&nbsp;    $SIG\{\_\_WARN\_\_\} = sub \{ 
&nbsp;    print "HERE IS THE WARNING&#8203;: $\_\[0\]";
&nbsp;    \# load the below module only when the subroutine is actually called
&nbsp;    require othermod;
&nbsp;    \};
\}

my $i;

\# Cause a compile warning
my $i;

\# END OF TESTCASE

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.6.0 - Thu Feb 15 18:59:18 PST 2001
It is being executed now by  Perl v5.6.1 - Thu Jul  5 10:26:43 PDT 2001.

Site configuration information for perl v5.6.1:

Configured by lakata at Thu Jul  5 10:26:43 PDT 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=solaris, osvers=2.6, archname=sun4-solaris
    uname='sunos spark 5.6 generic_105181-22 sun4u sparc '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.2 19991024 (release)', gccosandvers='solaris2.6'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib '
    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldl -lm -lc
    perllibs=-lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/sun4-solaris
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl/5.005/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.005
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/dj
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/lib:/usr/X/lib
    LOGDIR (unset)
    PATH=/mips/tools/synopsys/vcs/vcs6.0/sun/util:/mips/tools/synopsys/vcs/vcs6.0/sun/bin:/mips/tools/mips/dvp:.:/home/dj/scripts:/home/dj/bin:/usr/local/bin:/bin:/usr/bin:/usr/lib/X11:/usr/include/X11:/usr/adm/bin:/sbin:/usr/sbin:/mips/tools/lsf/bin:/usr/ucb:/usr/dt/bin:/usr/openwin/bin:/usr/local/frame/bin:/ess/proj/cherokee/dj/hardcore/build_script/bin:/ess/proj/cherokee/dj/hardcore/build_script/bin:/home/dj/l2/bin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Aug 07\, 2002 at 09&#8203;:36&#8203;:26PM \-0000\, Darren Jones wrote&#8203;:
> This is a bug report for perl from dj@&#8203;mips\.com\,
> generated with the help of perlbug 1\.28 running under perl v5\.6\.1\.

confirmed bug still present in 5\.8\.0

Dave\.

> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> \#\! /usr/local/bin/perl \-w
> \#
> \# The \-w flag is needed to cause the strange behavior\.
> \#
> \# This is a testcase for a very strange msg that perl is emitting at compile time&#8203;:
> \#
> \#   Can't use global @&#8203;\_ in "my" at othermod\.pm line 2\, near "return @&#8203;\_"
> \#   Compilation failed in require at djstrangefailure\.pl line 27\.
> \#
> \# There is no "my" declaration\, so I dont know exactly what is going wrong\.
> \# It seems to be confusing the parsing of the require block with that of the source code that caused the warning
> \#
> \# In order for this testcase to work\, you must have a file named othermod\.pm which contains
> \# the following 4 lines \(uncommented\, of course\!\)&#8203;:
> \#
> \# sub mysub \{
> \#    return @&#8203;\_;
> \# \}
> \# 1;
> \# 
> 
> \# Register a warning subroutine for compile warnings
> BEGIN \{
>     $SIG\{\_\_WARN\_\_\} = sub \{ 
>     print "HERE IS THE WARNING&#8203;: $\_\[0\]";
>     \# load the below module only when the subroutine is actually called
>     require othermod;
>     \};
> \}
> 
> my $i;
> 
> \# Cause a compile warning
> my $i;
> 
> \# END OF TESTCASE
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=medium
> \-\-\-
> This perlbug was built using Perl v5\.6\.0 \- Thu Feb 15 18&#8203;:59&#8203;:18 PST 2001
> It is being executed now by  Perl v5\.6\.1 \- Thu Jul  5 10&#8203;:26&#8203;:43 PDT 2001\.
> 
> Site configuration information for perl v5\.6\.1&#8203;:
> 
> Configured by lakata at Thu Jul  5 10&#8203;:26&#8203;:43 PDT 2001\.
> 
> Summary of my perl5 \(revision 5\.0 version 6 subversion 1\) configuration&#8203;:
>   Platform&#8203;:
>     osname=solaris\, osvers=2\.6\, archname=sun4\-solaris
>     uname='sunos spark 5\.6 generic\_105181\-22 sun4u sparc '
>     config\_args=''
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
>     useperlio=undef d\_sfio=undef uselargefiles=define usesocks=undef
>     use64bitint=undef use64bitall=undef uselongdouble=undef
>   Compiler&#8203;:
>     cc='gcc'\, ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>     optimize='\-O'\,
>     cppflags='\-fno\-strict\-aliasing \-I/usr/local/include'
>     ccversion=''\, gccversion='2\.95\.2 19991024 \(release\)'\, gccosandvers='solaris2\.6'
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>     alignbytes=8\, usemymalloc=y\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='gcc'\, ldflags =' \-L/usr/local/lib '
>     libpth=/usr/local/lib /usr/lib /usr/ccs/lib
>     libs=\-lsocket \-lnsl \-ldl \-lm \-lc
>     perllibs=\-lsocket \-lnsl \-ldl \-lm \-lc
>     libc=/lib/libc\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags='\-fPIC'\, lddlflags='\-G \-L/usr/local/lib'
> 
> Locally applied patches&#8203;:
>     
> 
> \-\-\-
> @&#8203;INC for perl v5\.6\.1&#8203;:
>     /usr/local/lib/perl5/5\.6\.1/sun4\-solaris
>     /usr/local/lib/perl5/5\.6\.1
>     /usr/local/lib/perl5/site\_perl/5\.6\.1/sun4\-solaris
>     /usr/local/lib/perl5/site\_perl/5\.6\.1
>     /usr/local/lib/perl5/site\_perl/5\.6\.0/sun4\-solaris
>     /usr/local/lib/perl5/site\_perl/5\.6\.0
>     /usr/local/lib/perl5/site\_perl/5\.005/sun4\-solaris
>     /usr/local/lib/perl5/site\_perl/5\.005
>     /usr/local/lib/perl5/site\_perl
>     \.
> 
> \-\-\-
> Environment for perl v5\.6\.1&#8203;:
>     HOME=/home/dj
>     LANG=C
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH=/usr/lib&#8203;:/usr/X/lib
>     LOGDIR \(unset\)
>     PATH=/mips/tools/synopsys/vcs/vcs6\.0/sun/util&#8203;:/mips/tools/synopsys/vcs/vcs6\.0/sun/bin&#8203;:/mips/tools/mips/dvp&#8203;:\.&#8203;:/home/dj/scripts&#8203;:/home/dj/bin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/lib/X11&#8203;:/usr/include/X11&#8203;:/usr/adm/bin&#8203;:/sbin&#8203;:/usr/sbin&#8203;:/mips/tools/lsf/bin&#8203;:/usr/ucb&#8203;:/usr/dt/bin&#8203;:/usr/openwin/bin&#8203;:/usr/local/frame/bin&#8203;:/ess/proj/cherokee/dj/hardcore/build\_script/bin&#8203;:/ess/proj/cherokee/dj/hardcore/build\_script/bin&#8203;:/home/dj/l2/bin
>     PERL\_BADLANG \(unset\)
>     SHELL=/usr/local/bin/tcsh
> 

\-\- 
Never do today what you can put off till tomorrow\.

**Comment by p5pRT** [comment]
### From @hvds

Darren Jones \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
:\# This is a testcase for a very strange msg that perl is emitting at compile time&#8203;:
:\#
:\#   Can't use global @&#8203;\_ in "my" at othermod\.pm line 2\, near "return @&#8203;\_"
:\#   Compilation failed in require at djstrangefailure\.pl line 27\.

This appears to be the same bug solved by Marty Pauley in change \#17928\,
so it should be fixed in future releases of perl\.

You'll find the analysis and patch at&#8203;:
&nbsp;  http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2002\-09/msg00820\.html

Hugo van der Sanden

**Comment by p5pRT** [comment]
### From dj@mips.com

Hi Hugo\,
Thanks for looking into this\. I do agree that the bug solved by Marty
Pauly is likely the same bug I am seeing\. I had one observation\, tho\-

In the Marty Pauly solution\, he seems to have changed the require'd
module's source \(utf8\.c\)\. The problem seems more internal to the Perl
kernel\- he is fixing in utf8\.c what should have been fixed inside perl
before passing control to utf8\.

In my problem\, All I have is perl code\- no C code to fix\. It seems to me
that perl should be clearing PL\_in\_my before compiling the require'd
module\.

I apologize if I am misunderstanding what is happening\, but I want to
make sure the bug is fixed\.

Darren

"Hugo van der Sanden \(via RT\)" wrote&#8203;:
> 
> Darren Jones \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
> :\# This is a testcase for a very strange msg that perl is emitting at compile time&#8203;:
> :\#
> :\#   Can't use global @&#8203;\_ in "my" at othermod\.pm line 2\, near "return @&#8203;\_"
> :\#   Compilation failed in require at djstrangefailure\.pl line 27\.
> 
> This appears to be the same bug solved by Marty Pauley in change \#17928\,
> so it should be fixed in future releases of perl\.
> 
> You'll find the analysis and patch at&#8203;:
>   http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2002\-09/msg00820\.html
> 
> Hugo van der Sanden

\-\- 
Darren Jones               Engineering Manager
MIPS Technologies          Phone&#8203;: \(650\)567\-5103
1225 Charleston Rd\.        mailto&#8203;:dj@&#8203;mips\.com
Mountain View\, CA 94043    http&#8203;://www\.mips\.com

**Comment by p5pRT** [comment]
perl.org@triv.org - Status changed from 'new' to 'resolved'
